File: clients/ios/Lextures/Core/LMS/LMSFeatureModelsCourseOutcomes.swift
import Foundation

/// Course outcomes settings models (M13.5).
struct CourseOutcomeLinkProgress: Codable, Hashable {
    var avgScorePercent: Double?
    var gradedLearners: Int
    var enrolledLearners: Int
}

struct CourseOutcomeLink: Codable, Identifiable, Hashable {
    var id: String
    var subOutcomeId: String?
    var structureItemId: String
    var targetKind: String
    var quizQuestionId: String
    var measurementLevel: String
    var intensityLevel: String
    var itemTitle: String
    var itemKind: String
    var progress: CourseOutcomeLinkProgress
}

struct CourseOutcome: Codable, Identifiable, Hashable {
    var id: String
    var title: String
    var description: String
    var sortOrder: Int
    var rollupAvgScorePercent: Double?
    var links: [CourseOutcomeLink]
}

struct CourseOutcomesListResponse: Codable, Hashable {
    var enrolledLearners: Int
    var outcomes: [CourseOutcome]
}

struct CreateCourseOutcomeBody: Encodable {
    var title: String
    var description: String
}

struct PatchCourseOutcomeBody: Encodable {
    var title: String?
    var description: String?
    var moduleStructureItemId: String?

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(title, forKey: .title)
        try container.encodeIfPresent(description, forKey: .description)
        try container.encodeIfPresent(moduleStructureItemId, forKey: .moduleStructureItemId)
    }

    private enum CodingKeys: String, CodingKey {
        case title
        case description
        case moduleStructureItemId
    }
}

struct AddCourseOutcomeLinkBody: Encodable {
    var structureItemId: String
    var targetKind: String
    var quizQuestionId: String?
    var measurementLevel: String
    var intensityLevel: String
    var subOutcomeId: String?
}

struct CourseOutcomeSubOutcome: Codable, Identifiable, Hashable {
    var id: String
    var outcomeId: String
    var title: String
    var description: String
    var sortOrder: Int
}

struct CreateCourseOutcomeSubOutcomeBody: Encodable {
    var title: String
    var description: String
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/LmsFeatureModelsCourseOutcomes.kt
package com.lextures.android.core.lms

import kotlinx.serialization.Serializable

@Serializable
data class CourseOutcomeLinkProgress(
    val avgScorePercent: Double? = null,
    val gradedLearners: Int = 0,
    val enrolledLearners: Int = 0,
)

@Serializable
data class CourseOutcomeLink(
    val id: String,
    val subOutcomeId: String? = null,
    val structureItemId: String,
    val targetKind: String,
    val quizQuestionId: String = "",
    val measurementLevel: String,
    val intensityLevel: String,
    val itemTitle: String,
    val itemKind: String,
    val progress: CourseOutcomeLinkProgress,
)

@Serializable
data class CourseOutcome(
    val id: String,
    val title: String,
    val description: String = "",
    val sortOrder: Int = 0,
    val rollupAvgScorePercent: Double? = null,
    val links: List<CourseOutcomeLink> = emptyList(),
)

@Serializable
data class CourseOutcomesListResponse(
    val enrolledLearners: Int = 0,
    val outcomes: List<CourseOutcome> = emptyList(),
)

@Serializable
data class CreateCourseOutcomeBody(
    val title: String,
    val description: String = "",
)

@Serializable
data class PatchCourseOutcomeBody(
    val title: String? = null,
    val description: String? = null,
)

@Serializable
data class AddCourseOutcomeLinkBody(
    val structureItemId: String,
    val targetKind: String,
    val quizQuestionId: String? = null,
    val measurementLevel: String,
    val intensityLevel: String,
    val subOutcomeId: String? = null,
)

@Serializable
data class CourseOutcomeSubOutcome(
    val id: String,
    val outcomeId: String,
    val title: String,
    val description: String = "",
    val sortOrder: Int = 0,
)

@Serializable
data class CreateCourseOutcomeSubOutcomeBody(
    val title: String,
    val description: String = "",
)